Nonconforming Kernel

int-ptr-ptr opened the pull request:
## Description

Please describe the changes/features in this pull request.
- adds integrate1d algorithm (test later)
- adds nonconforming to domain kernels
- nonconforming compute-coupling kernel

## Issue Number

If there is an issue created for these changes, link it here
#1265 

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

int-ptr-ptr committed:
additional accessors necessary for dg2d kernel

int-ptr-ptr committed:
removed `shmem_size` from point accessors

int-ptr-ptr committed:
isolated kernel from dg2d-kernelimpl

int-ptr-ptr committed:
src from kernelimpl, remove edge_to_interface

int-ptr-ptr committed:
add nc displacement test

int-ptr-ptr committed:
replaced get_transfer_function method with reference

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

int-ptr-ptr committed:
gpu not compiling again -- points to ncmarked_conforming_grid_flat

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

int-ptr-ptr committed:
modified cmakelists

Rohit-Kakodkar commented on include/algorithms/integrate/coupling_integral1d.hpp:
```diff
@@ -0,0 +1,104 @@
+                                          transfer_function_self);
+
+        PointFieldType result;
+#pragma unroll
```
These should be within `KOKKOS_ENABLE_CUDA` or atleast compiler ifdefs

Rohit-Kakodkar commented on include/algorithms/integrate/integrate1d.hpp:
```diff
@@ -0,0 +1,104 @@
+          ChunkEdgeWeightJacobianType::interface_tag,
+          ChunkEdgeWeightJacobianType::boundary_tag,
+          specfem::kokkos::DevScratchSpace,
+          Kokkos::MemoryTraits<Kokkos::Unmanaged>,
```
Point type should not need scratch space templates.

Rohit-Kakodkar commented on include/algorithms/integrate/integrate1d.hpp:
```diff
@@ -0,0 +1,104 @@
+          ChunkEdgeWeightJacobianType::boundary_tag,
+          specfem::kokkos::DevScratchSpace,
+          Kokkos::MemoryTraits<Kokkos::Unmanaged>,
+          IntersectionFieldViewType::using_simd>;
```
It would be hard for coupled interfaces to vectorized using simd since the data would not be contiguous in memory.

Rohit-Kakodkar commented on include/algorithms/integrate/integrate1d.hpp:
```diff
@@ -0,0 +1,104 @@
+    const IndexType &chunk_index,
+    const IntersectionFieldViewType &intersection_field,
+    const ChunkEdgeWeightJacobianType &weight_jacobian,
+    const CallableType &callback) {
```
- I suggest chaging the name for this function. I'd rather call it line_integral or something along those lines.

Rohit-Kakodkar commented on include/boundary_conditions/boundary_conditions.hpp:
```diff
@@ -45,6 +45,22 @@ apply_boundary_conditions(const PointBoundaryType &boundary,
   return;
 }
 
+// for some reason, the nonconforming kernel doesn't compile on cuda 11.8
```
Seems like for some reason its getting compiled. Which is weird since we have the same pattern for weakly conforming case.

Rohit-Kakodkar replied:
Do you know why this might be happening?

Rohit-Kakodkar committed:
Fix bug in coupled interfaces

- [x] Update index function in edge/intersection iterator
- [x] Update `for_all` execution function

int-ptr-ptr committed:
removed scratch constructor from point

int-ptr-ptr committed:
removed simd

int-ptr-ptr committed:
removed simd from test -- compiles now

Rohit-Kakodkar committed:
compiler fix

lsawade reviewed (approved):
I have nothing to add in addition to rohits comments

int-ptr-ptr committed:
compatibility with nci-accessor changes

int-ptr-ptr committed:
review changes

int-ptr-ptr committed:
polarized transfer algorithm, removed ambiguated transfer_function ref

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1350

int-ptr-ptr committed:
Merge remote-tracking branch 'public_repo/issue-1350' into nci-accessor

int-ptr-ptr committed:
ncmarked_conforming_grid_curved: rel src

revert  ncmarked_conforming_grid_curved CMakeLists.txt sources to relative (to match ncmarked_conforming_grid_flat CmakeLists)

int-ptr-ptr commented:
> Do you know why this might be happening?

Zero clue. Since it's only failing for CUDA 11.8 and not 12.8, it may be a bug with that version.
If you wanted to deprecate CUDA 11 support, then I believe we could remove those functions.

int-ptr-ptr committed:
Merge branch 'nci-accessor' into nci-kernel

int-ptr-ptr committed:
resolved post-merge compilation issues

int-ptr-ptr committed:
fixed newmark displacement test config

int-ptr-ptr committed:
bugfixes post-merge
